src/backend/distributed/utils/resource_lock.c

```diff
@@ -336,21 +338,58 @@ LockReferencedReferenceShardDistributionMetadata(uint64 shardId, LOCKMODE lock)
 
 	if (list_length(shardIntervalList) > 0 && ClusterHasKnownMetadataWorkers())
 	{
-		LockShardListMetadataOnWorkers(lock, shardIntervalList);
+		LockShardListMetadataOnWorkers(lockMode, shardIntervalList);
 	}
 
 	foreach(shardIntervalCell, shardIntervalList)
 	{
 		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
 
-		LockShardDistributionMetadata(shardInterval->shardId, lock);
+		LockShardDistributionMetadata(shardInterval->shardId, lockMode);
 	}
 }
 
 
 /*
- * GetSortedReferenceShards iterates through the given relation list.
- * Lists the shards of reference tables and returns the list after sorting.
+ * LockReferencedReferenceShardResources acquires resource locks with the
+ * given lock mode on the reference tables which has a foreign key from
+ * the given relation.
+ *
+ * It also gets resource locks on worker nodes to prevent concurrent write
+ * operations on reference tables from metadata nodes.
+ */
+static void
+LockReferencedReferenceShardResources(uint64 shardId, LOCKMODE lockMode)
+{
+	ListCell *shardIntervalCell = NULL;
+	Oid relationId = RelationIdForShard(shardId);
+
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+
+	/*
+	 * Note that referencedRelationsViaForeignKey contains transitively referenced
+	 * relations too.
+	 */
+	List *referencedRelationList = cacheEntry->referencedRelationsViaForeignKey;
+	List *shardIntervalList = GetSortedReferenceShardIntervals(referencedRelationList);
+
+	if (list_length(shardIntervalList) > 0 && ClusterHasKnownMetadataWorkers() &&
+		!IsFirstWorkerNode())
+	{
+		LockShardListResourcesOnFirstWorker(lockMode, shardIntervalList);
+	}
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		LockShardResource(shardInterval->shardId, lockMode);
+	}
```

Comment:
would it make sense to use `LockShardListResources` here or did you avoid it to skip the extra sort?

Reply:
> would it make sense to use `LockShardListResources` here or did you avoid it to skip the extra sort?

Yeah, just wanted to skip the extra sort.